hudi-common/src/main/java/org/apache/hudi/common/util/ReflectionUtils.java

```diff
@@ -48,19 +48,17 @@ public class ReflectionUtils {
   private static final Map<String, Class<?>> CLAZZ_CACHE = new HashMap<>();
 
   public static Class<?> getClass(String clazzName) {
-    if (!CLAZZ_CACHE.containsKey(clazzName)) {
-      synchronized (CLAZZ_CACHE) {
-        if (!CLAZZ_CACHE.containsKey(clazzName)) {
-          try {
-            Class<?> clazz = Class.forName(clazzName);
-            CLAZZ_CACHE.put(clazzName, clazz);
-          } catch (ClassNotFoundException e) {
-            throw new HoodieException("Unable to load class", e);
-          }
+    synchronized (CLAZZ_CACHE) {
```

Comment:
this requires you to take a lock on every call now, even when the value is in the cache. It will cause a hit in performance. What errors are you seeing in the containsKey operation before?

Reply:
https://github.com/apache/hudi/pull/9786 - is also looking at using concurrent hashmap which may simplify things

Reply:
@the-other-tim-brown get method in Map not use CAS for synchonized，this would cause data error such as npe，then add map::get in synchonized block

Reply:
`putIfAbsent` should already check the existence right?

Reply:
> `putIfAbsent` should already check the existence right?

@danny0405  from jdk1.8 source，putIfAbsent in HashMap/ConcuurentMap also seems not use CAS to ensure thread safe

Reply:
I mean why it is not safe here, we do not remove anything?

Reply:
"casTabAt(tab, i, null, new Node<K,V>(hash, key, value, null))" in ConcurrentMap::putIfAbsent，this should safe from CAS

Reply:
You mean the existence check does not have atomicity?

Reply:
> You mean the existence check does not have atomicity?

yes，so add the get to synchonized block @danny0405

Reply:
This is the first version of code, it is changed then by @the-other-tim-brown for performace issue.

Reply:
> This is the first version of code, it is changed then by @the-other-tim-brown for performace issue.

 yes，maybe we need revert it https://github.com/apache/hudi/pull/8433

Reply:
So there is a race condition between the `containsKey` check and the `get` call where the contains check is true but the value returned is null? Am I understanding that correctly? 

If this is the case, would changing the `containsKey` to a `get` call avoid this issue while avoiding taking a lock when possible.

Reply:
> So there is a race condition between the `containsKey` check and the `get` call where the contains check is true but the value returned is null? Am I understanding that correctly?
> 
> If this is the case, would changing the `containsKey` to a `get` call avoid this issue while avoiding taking a lock when possible.

@the-other-tim-brown you understand correctly，but in my option this not need the change as double if，like this pr is a better way

Reply:
This creates a bottleneck in the code since you need to take a lock every time you call into this method. This will block your threads even when the cache contains the class. Why is this better?

Reply:
> This creates a bottleneck in the code since you need to take a lock every time you call into this method. This will block your threads even when the cache contains the class. Why is this better?

sychnozied block in jdk8 had improve use biasedlock，this is not lock less and cost little，more important like this can avoid npe  and other thread safe problem. @the-other-tim-brown

Reply:
In case of 100 or more threads hitting this block, will we see the same performance as we have today? if yes, then I am fine with this change. I was just trying to avoid unnecessary cost when possible as we run a unique case with lots of Hudi writers/services hitting this block within a single JVM.

Reply:
> In case of 100 or more threads hitting this block, will we see the same performance as we have today? if yes, then I am fine with this change. I was just trying to avoid unnecessary cost when possible as we run a unique case with lots of Hudi writers/services hitting this block within a single JVM.
make sense，but if the block is a atomic obj like String、Integer this is very necessay，but map、set and other  collection may not fine in doube if way in synchonize block.

Reply:
> In case of 100 or more threads hitting this block, will we see the same performance as we have today? if yes, then I am fine with this change. I was just trying to avoid unnecessary cost when possible as we run a unique case with lots of Hudi writers/services hitting this block within a single JVM.

make sense，but if the block is a atomic obj like String、Integer this is very necessay，but map、set and other collection may not fine in doube if way in synchonize block.

Reply:
To provide some more insights, we're call this method on every record in the DeltaStreamer here: https://github.com/apache/hudi/blob/master/hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java#L627

That DataSourceUtils method is hitting this class cache so we should run a microbenchmark just to make sure this doesn't degrade our ingestion performance for users using Deltastreamer

Reply:
do you think we can keep the sync only for case where its not present. 
```
if (not present) {
   synchronized (OBJ) {
     instantiate 
     again check if its present and put only if absent.
    }
 }
```
So, most of the time regular readers who might end up reading an already existing object may not need to acquire lock